Attractive Sapphire Mantaray

High

# Borrow is repaid with wrong amount in _handleLiquidationSuccess.

### Summary

When the liquidation succeeds on Chain A and the LiquidationSuccess message is passed back to Chain B's _handleLiquidationSuccess, it calls repayCrossChainBorrowInternal to repay the borrower's debt on Chain B. However, the repayment amount `_amount` passed to repayCrossChainBorrowInternal is `payload.amount`, which is the number of collateral LTokens seized on Chain A (seizeTokens), instead of the `repayAmountFinal`.

When liquidateCrossChain occurs, function `_executeLiquidationCore` calculates the seizeTokens of lTokenToSeize:

```solidity
>@        (uint256 amountSeizeError, uint256 seizeTokens) = LendtrollerInterfaceV2(lendtroller)
            .liquidateCalculateSeizeTokens(borrowedlToken, params.lTokenToSeize, params.repayAmount);

        require(amountSeizeError == 0, "Seize calculation failed");

        // Send message to Chain A to execute the seize
        _send(
            params.srcEid,
>@            seizeTokens,
            params.storedBorrowIndex,
            0,
            params.borrower,
            lendStorage.crossChainLTokenMap(params.lTokenToSeize, params.srcEid), // Convert to Chain A version before sending
            msg.sender,
            params.borrowedAsset,
            ContractType.CrossChainLiquidationExecute
        );
```

Then in the function `_handleLiquidationExecute`, `payload.amount`(i.e., seizeTokens) is passed to the `_send`. 

```solidity
        _send(
            srcEid,
 >@           payload.amount,
            0,
            0,
            payload.sender,
            payload.destlToken,
            payload.liquidator,
            payload.srcToken,
            ContractType.LiquidationSuccess
        );
```

Finally, in the function _handleLiquidationSuccess, it's passed th the repayCrossChainBorrowInternal:

```soldity
repayCrossChainBorrowInternal(
            payload.sender, // The borrower
            payload.liquidator, // The liquidator (repayer)
>@            payload.amount, // Amount to repay
            payload.destlToken, // lToken representing the borrowed asset on this chain
            srcEid // The chain where the collateral (and borrow reference) is tracked
        );
```

### Root Cause

Function _handleLiquidationSuccess uses the error parameter to repay the borrow.

```solidity
        repayCrossChainBorrowInternal(
            payload.sender, // The borrower
            payload.liquidator, // The liquidator (repayer)
>@            payload.amount, // Amount to repay
            payload.destlToken, // lToken representing the borrowed asset on this chain
            srcEid // The chain where the collateral (and borrow reference) is tracked
        );
```

https://github.com/sherlock-audit/2025-05-lend-audit-contest/blob/main/Lend-V2/src/LayerZero/CrossChainRouter.sol#L443-L471

### Internal Pre-conditions

None

### External Pre-conditions

None

### Attack Path

None

### Impact

The repayment may fail because the payload.amount (i.e. the amount of collateral) is greater than the user's borrowed amount.

```solidity
    function repayCrossChainBorrowInternal(
        address borrower,
        address repayer,
        uint256 _amount,
        address _lToken,
        uint32 _srcEid
    ) internal {
        address _token = lendStorage.lTokenToUnderlying(_lToken);
        LTokenInterface(_lToken).accrueInterest();

        // Get borrow details and validate
        (uint256 borrowedAmount, uint256 index, LendStorage.Borrow memory borrowPosition) =
            _getBorrowDetails(borrower, _token, _lToken, _srcEid);

        // Calculate and validate repay amount
        uint256 repayAmountFinal = _amount == type(uint256).max ? borrowedAmount : _amount;
>@        require(repayAmountFinal <= borrowedAmount, "Repay amount exceeds borrow");
```

### PoC

_No response_

### Mitigation

Use the correct repayment amount `repayAmountFinal` instead of the collateral amount.